### ISSUE TITLE
Analyzer keeps treating a nullable property as is even after checking for that property to not be null

### ISSUE THREAD

**Issue by lsalazarm99** [dart-lang/language#1415]
Considering this code:

```dart
void main() {
  Person person = new Person(
    name: "Leonardo",
  );
  
  if (person.carBrand != null) {
    showInConsole(person.carBrand); // Error: The argument type 'String?' can't be assigned to the parameter type 'String'
  }
}

class Person {
    final String name;
    final String? carBrand;

    const Person({required this.name, this.carBrand});
}

void showInConsole(String text) {
  print(text);
}
```

In the error line, `person.carBrand` will not be null as it was checked for that before, but the analyzer keeps it with its original signature. Because of that, it is not usable inside the `showInConsole` function.

Some workarounds?

- Using a local variable for the property. In this case, I'm not sure if the analyzer keeps the local variable with its original signature after the null check, but it doesn't show an error anymore.

```dart
String? carBrand = person.carBrand;

if (carBrand != null) {
    showInConsole(carBrand); // no errors
}
````

- Using "!". But I'm not sure about this. Coming from TypeScript, this is a bad practice because if I remove the null check for some reason, it won't show the error and I will be throwing a runtime error.

```dart
if (person.carBrand != null) {
    showInConsole(person.carBrand!); // no errors
}
````

The first workaround is the safest way, I think. But what if I change the type of `cardBrand` in the future? I would need to change it in every location I check for the property to not be null. Also, it could be annoying.

If the analyzer can follow the flow of the first workaround for a local variable so it knows it's not null, shouldn't it do the same for the property of a local variable?

**Comment by tatumizer** [comment]
Please consider the following scenario:
```
class Person {
    final String name;
    final String? jobTitle;
    
    const Person({required this.name, this.jobTitle});
}
class PersonToBeDemoted extends Person {
  PersonToBeDemoted({required String name, required String jobTitle}):super(name: name, jobTitle: jobTitle);
  int requestCount=0;
  String? get jobTitle=> requestCount++==0? super.jobTitle: null;
}
main() {
  Person person = PersonToBeDemoted(name: "Eduardo", jobTitle: "Director");
  print(person.jobTitle); // now Eduardo is a Director
  print(person.jobTitle); // and now he is null
}
```
Things happen :-)

**Comment by lsalazarm99** [comment]
@tatumizer In that situation where you are getting the value from a function, I can understand that even if you check the return value to not be null, it doesn't mean the value won't be null when you run the function again (unless it is a pure function, but that might be more complex).

But what I am doing is not running a function so a non-ready-value. I am accessing a class property that is even marked as `final` so if I already checked for that property to not be `null`, it's impossible for that property to be `null`.

Also, I know `person` could be resigned so `jobTitle` could be `null` even after check. But if I mark `person` as `final`, the error is still present.

If I am checking the `final` property of a `final` variable to not be `null`, shouldn't it be enough for the analyzer to know it won't be `null` never?

**Comment by tatumizer** [comment]
Please take another look at my example.
In a subclass, the property `jobTitle` was **overridden by a getter**. "Finality" is lost! Because it's always possible to override a property (be it final or otherwise) with a getter, the compiler cannot use the logic based on "finality". (Note that the fact that the variable is declared as Person doesn't mean it's a basic person - it very well might be a PersonToBeDemoted)

**Comment by lsalazarm99** [comment]
I agree with you that, in your example, the error should be shown. As you said, `jobTitle` is overridden by a getter (so a function). And as a function, its return value might not be the same after every execution. That's why it should show an error.

But in my example, `jobTitle` is not a function. Is a final property. Yes, it can be overridden by another class, but I'm not using that other class. I'm using the class `Person` and its already `final` declared `jobTitle`. If I try to update the property or the `final` variable `person`, the analyzer won't let me because it will show an error telling me that it's not possible to set again those values.

```dart
void main() {
  final Person person = new Person(name: "Leonardo");
  
  person = new Person(name: "Eduardo"); // Error: The final variable 'person' can only be set once
  person.jobTitle = "Developer"; // Error: 'jobTitle' can't be used as a setter because it's final
  
  if (person.jobTitle != null) {
    // Since person can not be reassigned either its property jobTitle,
    // and I have already checked that person.jobTitle is not null,
    // it is safe to say that person.jobTitle won't be null never.

    showInConsole(person.jobTitle); // Error: The argument type 'String?' can't be assigned to the parameter type 'String'
  }
}

class Person {
    final String name;
    final String? jobTitle;

    const Person({required this.name, this.jobTitle});
}

void showInConsole(String text) {
  print(text);
}
```

Since those values can not be reassigned and the analyzer knows it, shouldn't it also know that if I'm checking for those values to not be null, those values won't be null inside the `if` statement?

In case those values could be reassigned, how could you reassign this so it is reasonable to show an error?

```dart
final Person person = new Person(name: "Leonardo");
```

**Comment by tatumizer** [comment]
> but I'm not using that other class

But how does the compiler know about that? It doesn't!
Well, you could argue that the compiler can see the declaration `Person person = new Person(name: "Leonardo");` and thus, the Person is just a basic Person, in which the jobTitle is final. But in the next version, you can modify it: `Person person = giveMeAPerson());` And now compiler cannot assume anything about the real type of Person and hence about the finality of jobTitle. It will show an error again. Do you see what I mean? The logic based on finality is very brittle. A small change in your program will break your program in ways you don't expect (the reason for the breakage is subtle). In other words, the logic based on finality would just create a pitfall that is better to be avoided. That's the justification for the current behavior.

**Comment by lsalazarm99** [comment]
> Well, you could argue that the compiler can see the declaration `Person person = new Person(name: "Leonardo");` and thus, the Person is just a basic Person, in which the jobTitle is final.

In this situation, checking for the property to not be null should be enough since, as you said, the compiler can see the declarations.

> But in the next version, you can modify it: Person person = giveMeAPerson()); And now compiler cannot assume anything about the real type of Person and hence about the finality of jobTitle.

In this situation, _where the code has changed_, since the compiler _doesn't know_ because I'm using a function, it is reasonable to show the null-safety error.

Maybe there is something I'm still not getting, but I'm not talking about the second situation. I'm talking about the first situation, which is a different one because the compiler can see that the property won't be null.

**Comment by srawlins** [comment]
>> but I'm not using that other class

> A small change in your program will break your program in ways you don't expect (the reason for the breakage is subtle).

I think this is an important consideration. I think the language team has looked at the various cases in which a developer says "but I _know_ the field isn't overridden," such as private fields (#1167), `@sealed` classes. Even in your case, the `Person()` constructor could be a factory constructor which returns a `PersonToBeDemoted`. Right now there are limited controls of variance in Dart; any time the static type of an object is `Person`, this means that the runtime type could be any subtype of `Person`.

One consolation is that the runtime behavior _might_ take knowledge into consideration with optimizations. Looking at:

```dart
if (person.jobTitle != null) {
    showInConsole(person.jobTitle!);
  }
```

If the compiler has optimizations (based on visibility, exact type, etc), then it may not need to execute that `!` null check. But for the sanity of the language, the `!` must remain because of specification of the static types.

**Comment by srawlins** [comment]
Duplicate of #1330

**Comment by tatumizer** [comment]
I think it would be good if the compiler, upon seeing
```
if (expr != null) {
   sameExpr.method;
} 
```
would not only complain about possible null but also point to an explanation of the phenomenon.
Otherwise, every user will be wondering why, filing issues, etc.

**Comment by Cat-sushi** [comment]
[label:field-promotion ](https://github.com/dart-lang/language/labels/field-promotion "Issues · dart-lang/language") is also related to this issue.

**Comment by tatumizer** [comment]
There are several parallel threads on the subject. It seems the issue will always be a source of confusion for users. Maybe the language team can take a step back and reconsider the tradeoffs?
Dart AFAIK always compiles the program from the sources - there's no partial compilation, precompiled libraries, etc. So the compiler always knows which field was overridden in a subclass, whether the constructor is a factory constructor and even then, what this factory constructor does, etc. The cases where the field can suddenly become null after the check for non-null are rare, but they are identifiable, and the error message can explain: so and so, you checked for non-null, but maybe it between such-and-such method changed the value underneath etc. In other words, compiler can, in principle, do something more "intelligent" than it does now. Is it a realistic idea?

**Comment by Cat-sushi** [comment]
If fields are promoted in some case and aren't promoted in other case, it is also confusing.

**Comment by tatumizer** [comment]
Can it be that the "other case" is so rare that the majority of users won't experience it in a lifetime?
But again, if it doesn't promote where the user expects it to promote, there should be a detailed explanation.

Neither solution is ideal. The question is just which of them constitutes the "least surprise".

**Comment by lrhn** [comment]
@tatumizer wrote:
> Dart AFAIK always compiles the program from the sources - there's no partial compilation, precompiled libraries, etc.

That is not correct. Dart does, at least for some compilers, have modular compilation. I believe the development compiler (dartdevc) is one. Among other things, that's what enables hot reload. If you had to wait for the entire program to be compiled, you'd at most have luke-warm reload. 😁 

Also, a package is a program part which is also compiled and run (for testing) independently of the program(s) it's part of.
If having a subclass changes the behavior of the superclass, then a superclass method getting field promotion due to not having a subclass, would break in a program where a subclass is added. 

It is vitally important to Dart's approach to object oriented programming (and ditto for most other languages with subclasses) that you can write a superclass without knowing its subclasses ahead of time. If you can write code in an instance method, which is then invalidated by a subclass, then you can't really use that feature safely *anywhere*. (Basically what @Cat-sushi said, if promotion inside a method depends on a global analysis of the entire final program, then it's incredibly fragile.)

Using global information for optimization is great, it enables more possible optimizations. 
Using global information to change semantics is not, it makes the semantics unpredictable and therefore unusable.

If you could *declare* that a class cannot have subclasses (or not subclasses outside of the declaring package), then you don't *need* global information (or seen differently, you *have* global information because the local declaration ensures global cooperation. And that only works with actual language features, not annotations that might or might not be checked).

**Comment by tatumizer** [comment]
>  if promotion inside a method depends on a global analysis of the entire final program, then it's incredibly fragile

Not sure if it is fragile or incredibly fragile, but whatever degree of fragility it is, current behavior doesn't make it do away. Consider the original example:
```
if (person.jobTitle != null) {
    showInConsole(person.jobTitle!); // no errors
}
```
Here. the user is encouraged to mindlessly add an exclamation mark. Even if the user does it mindfully, his mindfulness doesn't take into account the possibility of future changes in the program/libraries etc. (acting otherwise would be incredibly impossible). This exclamation mark stays in the program forever. My claim then is that the degree of fragility of this exclamation mark is exactly the same as the one you mentioned in your argument. The only difference is in the distribution of blame between the compiler and the programmer.

Can dart do better? E.g., the compiler doesn't complain (which would probably be OK in 99% of cases), but the linter (or another static analysis tool), either complains or doesn't depending on the global information about the program.
How about that?
(As you can see from the issue tracker, the status quo leads to widespread confusion).

**Comment by tatumizer** [comment]
I think an even stronger claim can be made: every instance of exclamation mark used in the program is as fragile as the one in the above example (and most of them are even "more fragile").
To justify the claim, let's look at the standard description of non-null assertion (I take it from Kotlin, but it's the same for every language):

**QUOTE**
## The !! Operator
The third option is for NPE-lovers: the not-null assertion operator (!!) converts any value to a non-null type and throws an exception if the value is null. We can write b!!, and this will return a non-null value of b (e.g., a String in our example) or throw an NPE if b is null:
```
val l = b!!.length
```
Thus, if you want an NPE, you can have it, but **you have to ask for it explicitly**, and it does not appear out of the blue.

**END QUOTE**

Sure, you have to ask for it explicitly. But who is "you"? Probably, it's the first person who put the non-null assertion in the code. This was done under specific circumstances where the person is confident in the safety of the assertion. But the circumstances may change. The "you" who put the assertion there is not the same "you" who makes the changes. Even if that is nominally the same person - it's a different version of the same person which may, or may no, remember about the assertion made a year ago. This is the same notion of fragility we are talking about. Clearly, it's not about fragility - it's all about the distribution of blame :-)

**Comment by lrhn** [comment]
It's about being able to *see* where things can fail. If you see `person.jobTitle!`, then you know that a `null` check occurs there. You can look to the surrounding code for hints to why the code feels safe to assume that the value won't be `null`.
And the author might consider whether there are better ways to do the same thing, one where you only make one type check.
That's why visibility is good, it allows you to notice things. It doesn't prevent you from doing something stupid if you really want to, but then again, nothing ever really does.

A `!` is still a run-time cast. It's a cast from `T?` to `T` (where the type system can usually figure out which non-nullable type you mean). It's still able to fail at run-time, just like `as T` would. We're definitely not trying to encourage anybody to spread `!` casts around their code. I hope that future code will more often be written in a way where it's not necessary to use `!`, or where it's used as sparingly as `as` should be now.

Existing code didn't have that possibility, and migrating existing code might (as the path of least rewrite) require adding `!` in places where the current code would also fail (possibly later) if the value is actually `null`.

**Comment by lsalazarm99** [comment]
Clearly you guys know more than me in this subject, and that's probably why I still can't understand something that **seems** obvious for me and probably many of the people with this issue.

`final Person person = Person(name: "Leonardo");`

In this sentence, I'm saying that the type of `person` is `Person`, and yes, it could be any subtype of `Person`. But then, through using `Person()` I'm saying that actually is not a subtype, but is just a `Person`.

So now the analyzer have all the information it needs (the exact type that the variable **is** and the information about that type).

Given that, shouldn't it know that if I'm checking for `person.jobTitle` to not be null, it won't be null?

Sure, in a future the code can change. If instead of using `Person()` I use `Developer()`, the analyser should then treat that variable as that subtype, but nothing else. If `Developer` overrides `jobTitle` with a getter, it will start complaining where you use your `person` variable. If it doesn't override it, it won't complain because it's still a simple nullable property that I already checked to not be null.

I can even change the `jobTitle` of `Person` and making it a getter and sure, I will understand it would need some refactor in my code because that's usually what it means to change the structure of a class.

But right know, in that line of code I wrote at the beginning (and the structure of `Person`), the analyzer have enough information to know that if I'm checking for `jobTitle` to not be null, it won't be null.

**Comment by tatumizer** [comment]
@lrhn: I am in total agreement with everything you said! In theory, you are 100% right!
Unfortunately, there are *theories* that are 100% right in theory and 100% wrong in practice, and this happens to be one of them (I wish it were the only theory of this kind, but it's not LOL).
Consider an example:
```
class Person {
  final String? jobTitle;
  Person([this.jobTitle]);
  void showInConsole(String s){}
  void foo() {
    if (jobTitle != null) {
      showInConsole(jobTitle); // error: The argument type "String?" cannot be assigned ...
    } 
  }  
}
```
The most likely reaction of the user is: 1) WTF? 2) filing an issue 3) adding an exclamation mark to `jobTitle`.
The important note here is that the scenario unfolds with respect to **every nullable instance variable**. (please correct me if I'm wrong).
So the fixed line reads: `showInConsole(jobTitle!)`.

Now, the important part: your argument "You can look to the surrounding code for hints to why the code feels safe to assume that the value won't be null." doesn't apply here. You can look to the surrounding code as much as you want, but you will never find anything for this `!` to make sense. From the surrounding code, you can see that the variable, to the best of your ability to reason, is guaranteed to be non-null inside "if" block. The naive user doesn't know that somebody else created a subclass that overrides "jobTitle" and made it not only non-final but quite volatile. (The very possibility of somebody overriding a final variable with a getter is far from being obvious.)

Please note that I'm not talking about the migration of old code. It's a totally new code. My claim is: every access to an *instance* variable must be decorated with an exclamation mark. At least, it would look like this for the typical user.
But the irony of the situation is that this "fix" is totally wrong! The only way to fix it correctly (which has nothing to do with an exclamation mark) is for the program to create a local copy:
```
    String? jobTitle = this.jobTitle;
    if (jobTitle != null) {
      showInConsole(jobTitle); 
    } 
 
```
And this is true for every nullable instance variable, on every access. 
But then, what about the "cheap" fix? 
```
    if (jobTitle != null) {
      showInConsole(jobTitle!);
    } 
``` 
Once we established that the code is fragile, it would stand to reason to expect the compiler to complain: so-and-so, this code is fragile, please create a local copy, etc. It's the same "fragility" argument as before.

Another point: contrary to expectations, the exclamation mark doesn't shout: look, I'm doing something potentially dangerous here! Because this happens so often (on each access to each instance variable) all exclamation marks cease to carry any useful information - they turn into annoying noise.
 
> I hope that future code will more often be written in a way where it's not necessary to use !, or where it's used as sparingly as as should be now.

I don't understand that. HOW? My code above is a new code. In this code, I HAVE TO either use `!`, which creates the fragility we were trying to avoid in the first place, or use a very hard-to-guess device with the local copy, which is never suggested by the compiler.

P.S. it's an interesting debate to have, I appreciate it :-)

**Comment by tatumizer** [comment]
@lsalazarm99:
> But right now, in that line of code I wrote at the beginning (and the structure of Person), the analyzer has enough information to know that if I'm checking for jobTitle to not be null, it won't be null.

Analyzer doesn't have this information. Without static analysis, it's not entirely clear even what the real type of Person is. Example:
```dart
class PersonToBeDemoted extends Person {
  PersonToBeDemoted({required String? jobTitle}):super._(jobTitle:jobTitle);
  int requestCount=0;
  String? get jobTitle=> requestCount++==0? super.jobTitle: null;
}
class Person {
  final String? jobTitle;
  Person._({this.jobTitle});
  factory Person({required String? jobTitle}) = PersonToBeDemoted;
}
```
When your program invokes Person("Director"), it doesn't get back an instance of basic Person - it always gets an instance of PersonToBeDemoted.

You can argue that the compiler can be smarter: it can look into the definition of Person and see whether it's a factory constructor or not, whether the name is final or not - it can do many things, but this exactly means that you want the compiler to do *static analysis*, which, as you can see from @lrhn's response, is costly and not always possible. All compiler knows is an interface of Person, with no implementation details.

I'm trying to argue that the compiler can be optimistic about *tested* variable:
```dart
if (instanceVar != null) {
   // compiler can assume instanceVar is not null here
}
```
but a separate static analysis tool can look deeper into the definitions. I simply can't see any way around it - see my previous post.

**Comment by lsalazarm99** [comment]
@tatumizer Many thanks, finally I understand what's happening and why is the shown error "valid".

**Comment by lrhn** [comment]
I generally create local variables if I have to check-and-use a nullable instance variable. That is a way to avoid the `!`, it's just a tad verbose.
I would very much like to have a feature which can check-and-bind in one operation, perhaps something like `if (x is Foo foo)` or `if (x is! Null foo)` which checks the value and binds it to a new variable *only* if it's not `null`. That allows me to introduce a new variable, check the value and promote the value along the successful check, which is the same thing I'm already doing, just with less verbosity.

I don't think we'll ever see sound promotion of virtual getters. We might be able to see it on non-virtual getters, but even there I'm somewhat against it because it would probably not work on *getters*, only fields, and that breaks the field/getter symmetry and locks classes into their current implementation.

Interfaces are promises about API and behavior, but only the API can really be checked by a computer. If we know that `x.foo` is a getter returning an `int`, that is *all* we know. We do not know that it's the same value each time you read it.
Dart has chosen to have fields and getters/setters be symmetric, so you can replace one with the other. That is, being a field is an implementation detail, not an API property. C# did something different, where they have getters and setters in the API, which are just methods called in special ways, but fields are completely different and (IIRC) not even virtual - so you have to write your getter/setter pair to access the field in a virtual way (like Java's `getFoo`/`setFoo`, just with nicer call syntax).
If we don't want to break that field/getter symmetry in Dart (and we still don't), we can't start assuming that even a final static field is promotable, because it would make it a breaking change to refactor that field into a getter.

Our problems here *are* largely self-inflicted. Because we can't look below the API to see that something really doesn't change (because either it's not guaranteed to be true because the getter is virtual, or because the promotion we'd do would then be fragile and depend on the getter staying a field), we can't promote anything except local variables.
I do not believe that can change without dropping some very fundamental assumptions in the Dart design.

So, I want to focus on providing a different and better way to get a promoted variable from any existing value, rather than focusing on linking separate reads of a fundamentally unknowable getter.

**Comment by rrousselGit** [comment]
Would it maybe make sense to introduce a syntax that prevents overriding a `final` variable with a getter?

**Comment by xvrh** [comment]
The proposal here: [field promotion with runtime checking](https://github.com/dart-lang/language/issues/1188) looks like the best solution. We trade a little bit of safety to A LOT MORE convenience.

All the examples showing why fields can't promote don't look realistic. In real codebase, 99% of fields/getters don't change type between 2 consecutive calls.

**Comment by rrousselGit** [comment]
I personally learn more toward sealed classes, preventing inheritance

If it is impossible to extends/implements a class, we know for sure that a `final` variable never changes, and the type-inference can work properly.

**Comment by lrhn** [comment]
Even sealed classes wouldn't take away the getter/field symmetry problem.
You'd still have to explicitly mark the field as unchanging, in a way where that can be strictly enforced by the compilers.

Say (obvious strawman): `final final int x = 42;`. This field cannot be overridden or implemented by anything but another `final final` declaration, so it's guaranteed that reading it twice will always give the same result. (And it's considered a breaking change to make a `final final` field into a mere `final` field, you're changing not just the implementation, but also the *promise* that the implementation won't change.

That would work even without sealed classes (you can implement the interface, but it better be with a `final final x` too!). It's an explicit promise to make a field always be unchanging. (Heck, we can even introduce `final get x => ...` getters that, somehow, promise to only ever have one value per instance, maybe by only returning the value of other "final" declarations. Starts to feel like `const` again, but just for the immutability).

**Comment by rrousselGit** [comment]
> Even sealed classes wouldn't take away the getter/field symmetry problem.

Why would that be a problem on a class that cannot be extended/implemented?

As it would not be possible to override a `final` with a getter. So there's never ever a case where a final property would change by reading it twice.

**Comment by tatumizer** [comment]
@lrhn:
> I generally create local variables if I have to check-and-use a nullable instance variable. That is a way to avoid the !, it's just a tad verbose.

The issue is not limited to nulls. Consider this program
```
class DoubleWrapper {
  final double value;
  DoubleWrapper(this.value){}
  double get reciprocal {
    if (value != 0) {
      return 1/value;  // FRAGILE? we are not guaranteed to get the same value!
    } else {
      throw "attempt to divide by zero";
    }
  } 
}
```
This is the same degree of fragility as we had with Person example, with the only difference that the compiler won't complain. 
Then why does it complain in the case of null check? In other words, why this gets flagged:
```
if (jobTitle != null) {
   showInConsole(jobTitle);
}
```
and this doesn't:
```
if (value != 0) {
   return 1/value;
}
```
If we consider one of them worthy of special treatment, then why not the other? So, if we want to avoid fragility, we can't use instance variables at all without copying into local variable, e.g.
```
// part of some method in a class
if (condition(x, y, z) { // x, y, z - instance vars
  use(x, y, z); // FRAGILE! x,y,z here may not be the same as in "condition"
}
```
I don't think "final final" can fix that. What if x, y, z are not final at all? Then you can't use them without copying?
This makes every "if" in your program a headache.

I think we are trying to solve a problem that is not worthy of solving. Compiler should accept this code without errors:
```
if (jobTitle != null) {
   showInConsole(jobTitle);
}
```
Sure, one can construct a contrived example where this program fails, but the probability of misfortune is low, and we generally don't buy insurance against unlikely events.

**Comment by Cat-sushi** [comment]
@tatumizer 
It is the problem that the current type system doesn't have double derived type which can't be 0, and it is not directly related to null safety.

**Comment by lrhn** [comment]
@rrousselGit Getter/field symmetry works both ways: A client of a class cannot see whether a property is implemented with a user written getter or an implicitly induced field getter (it's always a getter, sometimes it comes from a field), and the author is free to change between the two. It's an abstraction layer - the interface only tells you that you can read something, it doesn't promise how it's implemented.

If you can get promotion from a final field on a sealed class, but not from a getter on a sealed class, then the author of the class is no longer free to change the field to a getter. It has started mattering to clients how the getter is implemented.

The author might not know about this, after all, he might never see the code. So he could think that changing a final field to a getter on his already sealed class is a perfectly good refactoring. Instead it's a breaking change.

If that applies to *all* final fields of sealed classes, without any way to opt out of it, then a sealed class becomes *less* modifiable than a non-sealed class, which is the opposite of the reason to seal (to allow safely adding members later, because we know that no-one is implementing the class).

That is why I'm opposed to any attempt to derive unmodifiability of a getter from its choice of implementation &mdash; that will immediately make that implementation choice a part of the public API and break the deliberately introduced abstraction layer.
I want you to have to *opt in* to being "unmodifiable", so that it's a deliberate choice *and promise* that the implementation has certain properties. 

I think this is still a far more complicated approach than giving a way to promote-and-bind, which can be used for arbitrary values, rather than adding even more features to the object interface model, and then maintaining and preserving "unmodifiabilty". The bang is just not worth the bucks, they're better spent on something with more bang.

**Comment by tatumizer** [comment]
@Cat-sushi: you misunderstood my point. The problem has nothing to do with nulls, or doubles, or any concrete type.
The problem is that when you write
```
if (condition(x)) { // x - instance variable
   use(x); // x is not the same x that you checked on the previous line!
}
```
the value of "x" inside the block can be (in principle) different from the one you checked!

**Comment by Cat-sushi** [comment]
@xvrh 
If my understanding is correct,
in current Dart, even `late` variable have to be checked only once in every same context.
On the other hand, #1187 and #1188 are very expensive, because the compiler inserts null check everywhere referencing the fields.
So, I prefer #1191, #1201 or #1210.

**Comment by Cat-sushi** [comment]
@tatumizer 
I understand that It is a design decision that `x` might returns different value every time referenced.

**Comment by tatumizer** [comment]
> I understand that It is a design decision that x might returns different value every time referenced.

Yes, it's a design decision. Not an issue of type system. Not specific to nullable types. It's just for historical reasons, the problem became the focus of attention in the context of nullable types. My suggestion is to not do ANYTHING about it at all. As a trivial consequence of the (general!)  idea of doing nothing, the following code should be accepted:
```
if (x!=null) {
   showInConsole(x); // no compiler errors
}
```
Any attempt to do something special for nullable types in that regard will be inconsistent with the general case.

**Comment by Cat-sushi** [comment]
`print(x)` is valid regardless of `if(x == 5)`.
`showInConsole(x)` is invalid regardless of `if (x != null)`.
That is the difference.

**Comment by rrousselGit** [comment]
@lrhn 

> I want you to have to _opt in_ to being "unmodifiable", so that it's a deliberate choice _and promise_ that the implementation has certain properties.

But don't sealed classes qualify as an "opt in"? As the enhanced type-inference would be only available if the class is sealed, which requires an extra keyword on the class definition

In the end, the `final final` you're describing could be a subset of the "sealed" feature where:

```dart
sealed class Example {
  Example(this.a, this.b);

  final int a;
  final int b;
}
```

is equivalent to:

```dart
class Example {
  sealed Example(this.a, this.b);

  sealed final int a;
  sealed final int b;
}
```


As for getters, I believe `late` is the solution, where instead of:

```dart
int get c => a + b;
```

we can do:

```dart
late final c = a + b;
```

**Comment by tatumizer** [comment]
> print(x) is valid regardless of if(x == 5).

What do you mean by "valid"? Valid even if prints 7??

**Comment by Cat-sushi** [comment]
Syntactically correct.

**Comment by tatumizer** [comment]
@Cat-sushi: this is grammatically correct, too
```
if (x!=null) {
   showInConsole(x); // promotion, no compiler errors
}
```
If you mean type safety, the compiler can promote `x` here - the same way it promotes local variables.

**Comment by Cat-sushi** [comment]
No.
Overriding, field/getter symmetry prevent the promotion.

**Comment by tatumizer** [comment]
Here's my bet.
Suppose the compiler automatically transforms every statement like this:
```
if (x != null) {
   use(x);
}
```
into 
```
if (x != null) {
   use(x!); // ! is automatically inserted by a compiler.
}
```
The question is: how often will this transormation lead to unexpected NPE? Let's say, N lines out of each 1 million lines of **real** (not imaginary) code from pub (on average)
My bet is: N < 1
Why am I so sure? Because the problem is not specific to nullable types. It is not a new problem at all. If the tricks like PersonToBeDemoted were common, then nothing would ever work in existing dart code from sdk (see the example with x == 5 above - it has nothing to do with nullable types).
Look at the issue tracker - the latest 4 entries are about non-null promotion. This is the showstopper. All the cures suggested so far are worse than the disease: they add verbosity without adding any tangible value IMO. Doing too much about non-existent problem will serve no practical purpose but damage a good language.

**Comment by Cat-sushi** [comment]
@tatumizer
duplicate of #1188

**Comment by lrhn** [comment]
@rrousselGit 
> But don't sealed classes qualify as an "opt in"?

I don't see how. You're opting in to not *extending* or *implementing* the interface. If that *also* means not being able to change a field to a getter, then you'd be opting in to two things at the same time, whether or not you only wanted one of them.
You would have no way to define a sealed interface with a field (getter) that you might change to a getter in the future.

Linking the two would be counter-productive. You would only be able to have "final final" fields if you seal, and you won't be able to have mere "final" fields if you do seal.

Using `sealed final ...` instead of `final final ...` is probably easier on the eyes, but slightly different in meaning to me.

@tatumizer 
It's definitely possible to make `if (instanceVar != null) use(instanceVar)` automatically insert a `!` (an implicit downcast). 
It's just not the way we have otherwise moved with Dart, now that we have otherwise removed implicit downcasts (from anything but `dynamic`). The people clamoring for no implicit downcasts would not be happy about re-introducing them just to avoid writing `!`.
And it won't change the meaning at all, compared to if you just write the `!`. It's still a potentially failing run-time check.
For now, we have chosen to make such potentially failing run-time checks explicit. That's why we introduced `!` to begin with, so that it's still brief.

**Comment by rrousselGit** [comment]
> You would only be able to have "final final" fields if you seal, and you won't be able to have mere "final" fields if you do seal.

Getters wouldn't be "sealed"

So you can do:

```dart
sealed class Example {
  Example(this._a);

  final int _a;
  int get a => _a;
}
```

**Comment by esDotDev** [comment]
This really needs a solution now that nnbd is rolling into flutter at large. 

Without one, we're going to end up with a bunch of landmines like this:
```dart
void _handlePastePressed() async {
    if (controller == null) return;
    int start = controller!.selection.start;
    removeTextRange(controller!.selection.start, controller!.selection.end);
    String? value = (await Clipboard.getData("text/plain"))?.text;
    if (value != null) {
      addTextAtOffset(controller!.selection.start, value);
      // Move cursor to end on paste, as one does on desktop :)
      controller!.selection = TextSelection.fromPosition(TextPosition(offset: start + value.length));
    }
  }
```
This is not safe code at all, if a developer accidentally removes the null check at the top, the compiler would say nothing at all, and this is just a RTE waiting to happen. 

The right way to code this would be to add somethng like 9 `?` characters and several inline `?? 0`, or convert controller to a local var, and do the nulll check on that. But both are clumsy and verbose.

My worry is that devs will opt for the ! here, since it's by far the least invasive and easiest to read, but they are really filling their methods with potential landmines.

It seems sad that such a code-smelly feature like overriding final fields with accessors, is forcing such a big weakness in something as core as sound null safety.

**Comment by esDotDev** [comment]
The only way I can see this happening Flutter is this bizarre concoction: 
```
class Foo extends StatelessWidget {
  const Foo({Key? key, this.value}) : super(key: key);
  final String? value;

  @override
  Widget build(BuildContext context) {
    return (value == null) ? Container() : Text(value);
  }
}

class Bar extends Foo {
  @override
  String? get value => null;
}
```

Intuitively, I'm sure most developers would be shocked to learn you can override a final field, with an accessor without a peep from the compiler, and this is a source of most of the confusion. No one knows you can do this, and no one does this, cause it's bizarre and probably an anti-pattern out of the gate.

In C# the compiler would warn you to make this accessor as 'new' because it's replacing the base value and this should be called out to anyone reading. And if you tried to mark it as override, it would not allow it at all:
![image](https://user-images.githubusercontent.com/736973/110582545-9b0e9b00-8129-11eb-9adf-51d642ed6499.png)

This is along the lines of what most developers would expect I think.

**Comment by lrhn** [comment]
Overriding a field with getters/setters makes sense for *intercepting* more than it does for *replacing*.

Overriding a field (final or not) with a getter, and totally shadowing the field, is an anti-pattern because it's a waste of memory.
A much more reasonable pattern would be a getter which forwards to the super-class field, but does something else on the side (perhaps logging access), or even more reasonably, having a setter which forwards the value to the field, but also reports the change like an observable.

Fields in Dart are *always* virtual, unlike in C# and Java.
That's why we don't need an extra `getFoo`/`setFoo` virtual method pair like in Java. 
C# has a syntax shorthand for it, `public int Foo {get; set;}` which creates a field with a fresh name, and default virtual get/set operations backed by it. It's like Dart *only* has that syntax, there are no non-virtual fields.

**Comment by esDotDev** [comment]
This seems like it would work ok? The issue is not the ! operators themselves, but rather that they can be 
"orphaned" if the preceding null check is ever removed, so by having the compiler do this manually we sidestep the issue?

**Comment by tatumizer** [comment]
No, the issue is ! operators themselves :-)
Compiler cannot silently insert an exclamation mark simply because it's logically wrong. It's very easy to provide a counterexample - the bug report against this feature will be filed on day one. Who knows how this bug can be exploited, but good people will find a way :-)

**Comment by eernstg** [comment]
About this idea:

> Here's my bet.
> Suppose the compiler automatically transforms every statement like this:
> 
> ```
> if (x != null) {
>    use(x);
> }
> ```
> 
> into
> 
> ```
> if (x != null) {
>    use(x!); // ! is automatically inserted by a compiler.
> }
> ```

That's https://github.com/dart-lang/language/issues/1188.

**Comment by esDotDev** [comment]
> No, the issue is exactly ! operators themselves :-)
> Compiler cannot silently insert an exclamation mark simply because it's logically wrong. It's very easy to provide a counterexample - the bug report against this feature will be filed on day one. Who knows how this bug can be exploited, but good people will find a way :-)

Lol, fair point! But I am speaking purely from a pragmatic standpoint, I do not mind placing ! operators, or having them exist after an explicit null-check, none of that makes my app code more brittle.

What I do mind, is putting in bang operators, that are semantically weakly linked to some other line in a method, that can potentially landmines down the road. We need some way to say, if this null check was removed, the compiler should light up again.

Totally agree, this really should not be necessary at all, but also don't want to wait 3 yrs for a perfect solution, when a good enough will do.

**Comment by cedvdb** [comment]
>  We need some way to say, if this null check was removed, the compiler should light up again.

Isn't that exactly what it would do ? If the null check is removed, the ! isn't inserted and the compiler will light up.. This discussion is very bizar. It's counter intuitive to use ! after a null check. If a getter overrides that null check as someone proposed, you cannot even know, because you are not supposed to go into the source code after a null check.. If you take that at face value it makes the null check an anti pattern which it obviously isn't. At the code the null check is run upon is just not robust enough and there is nothing the language can do against, so far.

**Comment by esDotDev** [comment]
> How do you even work with such variables

You must read them once and cache the result locally.

Most seem to agree that this is a flaw in the language design that this is necessary, but also that the ship has sailed and it is what it is.

So there's various proposals to help mitigate this, I think the leading one is probably if-vars, which would be syntactic sugar for local var shadowing.
```
if(var jobTitle != null){
  print(jobTitle.length); // jobTitle is now a non-null string
  jobTitle = "NewTitle"; // this gets copied back to class field
}
```
https://github.com/dart-lang/language/issues/1201

**Comment by cedvdb** [comment]
> > How do you even work with such variables
> 
> You must read them once and cache the result locally.

I thought about that but is that really the way people code? This hole exists in many languages and again the hole is more in the underlying implementation of the getter than the language itself..

**Comment by esDotDev** [comment]
> I thought about that but is that really a language you'd have fun using?

You're preaching to the choir here, I totally agree. No one writes code this pedantically by hand, I'm not sure why the compiler needs to be so strict here, especially when that strictness leads directly to !'s shooting your null safety in the foot.

**Comment by danielgomezrico** [comment]
This issue makes the code noisy, it makes:
- Adding all the `!` where it makes no sense adding noise
- Devs gets used to add `!` everywhere and that may cause issues where it should not exist
- Devs that are new to `nullable types` get confused about the concept, and to me that is pretty important.

I remember that this was an issue on Kotlin, and they fixed it, is there any possibility to add some priority to this?

**Comment by mateusfccp** [comment]
@danielgomezrico 

I suppose that you've read the discussion above that provides the information regarding why this is this way and why it's not simple to solve it, at least not in a non-breaking way.

There are many proposals that intends to propose this issue completely or partially, you may have a look at them by the label `field-promotion`.

Although I agree that this issue is a little annoying, your three problems may be simply solved by instructing your team. For example, in my team we instruct our developers to never use `!`, and use promotion instead. PRs that use `!` instead of promoting the values are not accepted.

**Comment by caseycrogers** [comment]
Has this discussion changes at all with the introduction of `final`? From what I can tell, the key problem now that `final` exists is as follows (copied form @lrhn earlier):

>Getter/field symmetry works both ways: A client of a class cannot see whether a property is implemented with a user written getter or an implicitly induced field getter (it's always a getter, sometimes it comes from a field), and the author is free to change between the two. It's an abstraction layer - the interface only tells you that you can read something, it doesn't promise how it's implemented.

Yes, I agree, it'd be pretty non-ideal if using the `final` class keyword implicitly made field->getter migrations breaking changes. But, having to constantly create local variables just to get null inference is also a huge pain.

Flutter's `AsyncSnapshot` is a good example of this. I have to check `data` for nullability twice and/or unpack it into a local variable. I've seen a lot of codebases that have their own bespoke equivalent to `AsyncSnapshot` too that adds bells and whistles and every time the tension between redundant and runtime-brittle `!` checks vs verbose unpacking is a pretty deeply felt pain.
See the Flutter `FutureBuilder` example which chooses the brittle path (display the `data` object, assuming at runtime that it is not null):
https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html
The "check null, then use `!~" pattern pattern is so common that `AsyncSnapshot` even provides a convenience wrapper for it with `requireData`. FWIW, the ship has probably sailed on `AsyncSnapshot` as even if we could get null inference on attributes of final classes, making `AsyncSnapshot` final would probably be far too much of a breaking change. But given that most projects (including the one I'm working on right now) use bespoke alternatives anyway, those could at least migrate to `final`.

So we have to weigh the tradeoff here. Personally, I think improved convenience and/or safety for null checking attributes is way more important than the API purity and marginal reduction in potential breaking changes, but I understand if the Dart team still comes down on the other side.

**Comment by rrousselGit** [comment]
Quite a lot has changed in two years.

In 2023, we got pattern matching, which empowers `if` to define variables:

```dart
// Equivalent to `final field = obj.field; if (field != null)`
if (obj.field case final field?) {
  field.doSomething(); // OK
}
```

 And promotion for private fields:

```dart
if (obj._value != null) {
  obj._value.toDouble(); // OK
}
```


Ideally promotion would work for public fields too. But I doubt that this is realistic, given how Dart works.

To me, this issue is solved.

**Comment by caseycrogers** [comment]
I don't think the pattern matching really helps at all because you're almost always using an early return pattern where you return if null. The alternative-checking if not null and putting the body in the if brackets-tends to be untenable because you end up with unreadable amounts of nesting.

Private fields is nice, but also doesn't easily help in the majority of scenarios.

The current place where this most annoys me is with hooks-the code base is filled with hooks that construct and return a record that includes all nullables like errors, data and "refresh/set" callbacks. Records were used instead of classes because the team considered null type promotion a non negotiable requirement. But now the type signatures and general contract for the hooks are really obtuse.

My two cents is that the experience has only improved for a very limited slice of scenarios.

**Comment by rrousselGit** [comment]
Early returns are a fair shout out.

I wonder if we could have a variant to `if case` but with early returns in mind.

For example:

```dart
// !case is the reverse of case
if (obj.field !case final field?) return;

field.doSomething();
```

**Comment by mateusfccp** [comment]
> I don't think the pattern matching really helps at all because you're almost always using an early return pattern where you return if null. The alternative-checking if not null and putting the body in the if brackets-tends to be untenable because you end up with unreadable amounts of nesting.

I often do this, and it's not less readable (I actually think it's more readable than early returns, to have all returns in the same nest level). Most of the time (like 90% at least in my personal experience), you will have only a single level of nesting, not a nesting hell.

This is a subjective experience, though.

**Comment by tatumizer** [comment]
```dart
if (obj.field !case final field?) return;

field.doSomething();
```
This is *logically* impossible to implement if you allow variables introduced in `case` clause to shadow those from the outer context.

**Comment by rrousselGit** [comment]
> ```dart
> if (obj.field !case final field?) return;
> 
> field.doSomething();
> ```
> 
>   
> This is logically impossible to implement if you allow variables introduced in case clause to shadow those from the outer context.


I'm not sure about that, but I'm not attached to this exact syntax.

This feels reminiscent of the `guard` discussions.
We could imagine something else, like:

```dart
void fn(MyObject obj) {
  guard final field? = obj.field;
  else
    return;
}
```

Or whatever the syntax needs to be.
That `else return;` could even be optional in `void` functions.

**Comment by tatumizer** [comment]
Quote:
In guard or if let statements, you cannot shadow a constant or variable declared in the outer scope using the same name.
```swift
let x = 10
guard let x = someOptional else {  // Error: Cannot assign to value of 'x' because it is already declared.
    return
}
```
End Quote
(source: chatGPT)

In dart, there's no such restriction

**Comment by caseycrogers** [comment]
Can we get null promotion for nullable fields within a final class? Sort of like we already have with private methods? I think this unlocks 90% of the value (in the Flutter context) without hitting any of the blockers that have come up in this thread so far.
Credit to @LukePighetti for the idea.

Namely:
- We already don't abstract everything to a getter from within the class (see private attributes getting null promoted)
- If the `final` null promotion only applies within-class, then field -> getter migrations still aren't breaking changes because the only places where you got the stronger null promotion were within the class itself

This would look like this:
```dart
// FInal isn't a problem because you're rarely ever subclassing widgets anyway.
final class FooView extends StatelessWidget {
  const FooView({this.foo});

  final Foo? foo;

  @override
  Widget build(BuildContext context) {
    if (foo == null) return SizedBox.shrink();

    // No `!` needed because from within a final class we would get null promotion on fields.
    return Text(foo.bar);
  }
}

// In some other file...
final fooView = tester.widget<FooView>(find.byType(FooView));
// Doesn't work. Consumers outside of the class/file see `foo` abstracted as a getter
// so they can't get null promotion.
if (fooView.foo != null) print(fooView.foo.bar);
```

**Comment by lrhn** [comment]
> Can we get null promotion for nullable fields within a final class?

It's possible. It does require that all subclasses inside the same library (if any) are also `final` or `sealed`, and declares the property as a final instance variable too. The promotion only applies inside that library. (Doesn't have to be just the class.)

**Comment by caseycrogers** [comment]
@lrhn coming back to this, at Betterment this is one of the limitations of Dart that hurts dev ex the most for us. We have A LOT of named nullable constructor arguments and our build functions are littered with scenarios like:
```dart
class FooView extends StatelessWidget {
    ...
    final effectiveFoo = foo;
    if (foo == null) {
      return SizedBox.shrink();
    }
    return Text(foo);
  }
}
```

We never really have a reason to sub-class `FooView` so declaring it final would be fine and would let us cut out the `effectiveFoo` which is confusing and pollutes the namespace in the build function (especially because in practice we often have 3-4 nullable named args in a single widget, not just one).

I surveyed our Flutter devs for the Dart and Flutter dev ex issues they cared about and had them vote on which one they most wanted fixed and this issue got the most votes-hence why I'm coming back to this now. Does any of this background bump the priority of this issue for you all?
(also happy to share the details of our internal dev survey in a private channel if you're interested)

**Comment by lrhn** [comment]
You can use patterns to bind a tested value:
```dart
if (foo case final foo?) {
  return Text(foo.bar);
}
return SizedBox.shrink();
```
Works best if the promoted branch comes first.

Or it can use a switch:
```dart
return switch (foo) {
  null => return SizedBox.shrink(),
  var foo => return Text(foo.bar),
}
```
or the last line could even be:
```dart
  Foo(:var bar) => return Text(bar),
```

**Comment by esDotDev** [comment]
The cure is worse than the disease here. You save a line, but the result is overwrought and unreadable.

**Comment by caseycrogers** [comment]
We could make the fields private, but they're all named arguments (they're widgets, so positional arguments are almost never used) and declaring private attributes for named arguments is super tedious. This was actually my first intuition for the right solution here. Here's my comment where I outlined different possible approaches to improve Dev Ex for named constructor arguments for private fields:
https://github.com/dart-lang/language/issues/2005#issuecomment-2669551394
If I could pick between getting (library-scoped) null promotion on public attributes in final classes vs getting better syntax for named arguments for private fields, I'd much prefer the latter and my null promotion needs would be largely met.

For your comment on complexity/verbosity, my issue with the `effectiveFoo` pattern is not the extra line and much more than it confuses the namespace and adds an extra layer of indirection (more clickthroughs to trace a code path, using IDE to rename doesn't propagate, etc).

I've found (and fixed) code like the following (not actually a user facing bug so not high stakes, but illustrates why having duplicate references to the same variable in the namespace is confusing):
```dart
  final effectiveFoo = this.foo;
  
  if (effectiveFoo == null) return SixedBox.shrink();

  return Column(
    children: [
      SomeOtherWidget(),
      // If check always passes, should just use null promoted `effectiveFoo`.
      if (foo case final nonNullFoo?) FooView(foo);
    ]
  );
```

I could reuse the variable name and willfully shadow, but that comes with its own confusion:
`final foo = this.foo;`

PS: I hope I don't come off like I'm being willfully difficult here, there are just A LOT of ways to approach this developer need and over the last couple quarters I've put a lot of thought into the different approaches and it's hard to describe upfront off the top of my head all the different constraints/factors here-I'm finding this example/counterexample exercise helpful and hope it is helpful for your all as well
